File: java/ZigZag_Conversion.java
// use the strategy of cycle
// time is O(n) space is O(n)
public class Solution {
    public String convert(String s, int numRows) {
        StringBuffer result = new StringBuffer() ;
        int cycle = numRows + numRows -2;
        if ( cycle == 0) return s ;
        for (int i = 0 ; i < numRows ; ++i) {
            if (i != 0 && i != numRows - 1) {
                for (int j = i, k = cycle - i ; j < s.length() || k < s.length() ; j += cycle, k += cycle) {
                    if (j < s.length()) {
                        result.append(s.charAt(j)) ;
                    }
                    if (k < s.length() ) {
                        result.append(s.charAt(k)) ;
                    }
                    
                }
            }
            else {
                for (int j = i; j < s.length() ; j += cycle){
                    result.append(s.charAt(j)) ;
                }
            }
        }
        
        return result.toString();
    }
}

File: Python/ZigZag Conversion.py
class Solution:
    # @param {string} s
    # @param {integer} numRows
    # @return {string}
    def convert(self, s, numRows):
        str = [[] for i in range(0, numRows)]
        pt  = 0
        add = 0
        for i in range(0,len(s)):
            str[pt].append(s[i])
            if pt == numRows - 1 and pt != 0:
                add = -1
            elif pt == 0 and pt != numRows - 1:
                add = 1
            pt += add
        return "".join(["".join(str[i]) for i in range(0, numRows)])
